source/fuzz/added_function_reducer.h

```diff
@@ -0,0 +1,189 @@
+  // function to make it livesafe).  This function has |added_function_id_| as
+  // its result id.
+  //
+  // The ids associated with all global variables in |binary_out| that had the
+  // "irrelevant pointee value" fact are also returned via
+  // |irrelevant_pointee_global_variables|.
+  //
+  // The point of this function is that spirv-reduce can subsequently be applied
+  // to function |added_function_id_| in |binary_out|.  By construction,
+  // |added_function_id_| should originally manipulate globals for which
+  // "irrelevant pointee value" facts hold.  The set
+  // |irrelevant_pointee_global_variables| can be used to force spirv-reduce
+  // to preserve this, to avoid the reduced function ending up manipulating
+  // other global variables of the SPIR-V module, potentially changing their
+  // value and thus changing the semantics of the module.
+  void ReplayPrefixAndAddFunction(
+      std::vector<uint32_t>* binary_out,
+      std::unordered_set<uint32_t>* irrelevant_pointee_global_variables) const;
+
+  // This is the interestingness function that will be used by spirv-reduce
+  // when shrinking the added function.
+  //
+  // For |binary_under_reduction| to be deemed interesting, the following
+  // conditions must hold:
+  // - The function with id |added_function_id_| in |binary_under_reduction|
+  //   must only reference global variables in
+  //   |irrelevant_pointee_global_variables|.  This avoids the reduced function
+  //   changing the semantics of the original SPIR-V module.
+  // - It must be possible to successfully replay the transformations in
+  //   |transformation_sequence_in_|, adapted so that the function added by the
+  //   transformation at |index_of_add_function_transformation_| is replaced by
+  //   the function with id |added_function_id_| in |binary_under_reduction|,
+  //   to |binary_in| (starting with initial facts |initial_facts_|).
+  // - All the transformations in this sequence must be successfully applied
+  //   during replay.
+  // - The resulting binary must be interesting according to
+  //   |shrinker_interestingness_function_|.
+  bool InterestingnessFunctionForReducingAddedFunction(
+      const std::vector<uint32_t>& binary_under_reduction,
+      const std::unordered_set<uint32_t>& irrelevant_pointee_global_variables);
+
+  // Starting with |binary_in_| and |initial_facts_|, the transformations in
+  // |transformation_sequence_in_| are replayed.  However, the transforamtion
+  // at index |index_of_add_function_transformation_| of
+  // |transformation_sequence_in_| -- which is guaranteed to be an AddFunction
+  // transformation -- is adapted so that the function to be added is replaced
+  // with the function in |binary_under_reduction| with id |added_function_id_|.
+  void ReplayAdaptedTransformations(
+      const std::vector<uint32_t>& binary_under_reduction,
+      std::vector<uint32_t>* binary_out,
+      protobufs::TransformationSequence* transformation_sequence_out) const;
```

Comment:
Mention |binary_out| and |transformation_sequence_out| ?

Reply:
Done.